http-body/src/frame.rs

```diff
@@ -0,0 +1,68 @@
+use http::HeaderMap;
+
+/// A frame of any kind related to an HTTP stream (body).
+#[derive(Debug)]
+pub struct Frame<T> {
+    kind: Kind<T>,
+}
+
+#[derive(Debug)]
+enum Kind<T> {
+    // The first two variants are "inlined" since they are undoubtedly
+    // the most common. This saves us from having to allocate a
+    // boxed trait object for them.
+    Data(T),
+    Trailers(HeaderMap),
+    //Unknown(Box<dyn Frameish>),
+}
+
+impl<T> Frame<T> {
+    /// Create a DATA frame with the provided `Buf`.
+    pub fn data(buf: T) -> Self {
+        Self {
+            kind: Kind::Data(buf),
+        }
+    }
+
+    /// Create a trailers frame.
+    pub fn trailers(map: HeaderMap) -> Self {
+        Self {
+            kind: Kind::Trailers(map),
+        }
+    }
+
+    /// Returns whether this is a DATA frame.
+    pub fn is_data(&self) -> bool {
+        matches!(self.kind, Kind::Data(..))
+    }
+
+    /// Returns whether this is a trailers frame.
+    pub fn is_trailers(&self) -> bool {
+        matches!(self.kind, Kind::Trailers(..))
+    }
+
+    /// Consumes self into the buf of the DATA frame.
+    ///
+    /// # Panics
+    ///
+    /// If this isn't a DATA frame, this will panic. You should check `is_data`
+    /// before calling this.
+    // === REVIEW: is this the right method to expose? ===
+    pub fn into_data(self) -> T {
+        match self.kind {
+            Kind::Data(data) => data,
+            _ => panic!("called into_data on a non-data frame"),
+        }
+    }
+
+    /// If this is a DATA frame, returns a reference to it.
+    ///
+    /// Returns `None` if not a DATA frame.
+    // === REVIEW: is this the right method to expose? ===
```

Comment:
I imagine there being "easy" accessors for the 2 most common frames. Should they be `as_data()`/`as_data_mut()`, or `data_ref()`/`data_mut()`, or something else that's even better?

Reply:
I think `data_ref()`/`data_mut` is better https://doc.rust-lang.org/std/path/struct.Path.html#method.extension since you're not really converting a type but exposing an inner detail.

Reply:
It's kinda like `String::as_str()`, but only if the internal type is `data` (or `trailers`, etc).